### ISSUE TITLE
0.10.6: type attribute not respected on belongs to relations

### ISSUE THREAD

**Issue by rmachielse** [rails-api/active_model_serializers#2125]
When using a custom `type` on a serializer, the type is not used on nested belongs to relations.
This only happens with belongs_to relationships, has_many relationships are working fine.

```ruby
class Foo < ActiveRecord::Base
end

class Bar < ActiveRecord::Base
  belongs_to :foo
end

class FooSerializer < ActiveModel::Serializer
  type 'custom-type'
end

class BarSerializer < ActiveModel::Serializer
end
```

#### Expected behavior vs actual behavior

Expected:

```json
{ "data": { "relationships": { "foo": { "data": { "id": 1, "type": "custom-type" } } } }
```

Actual:

```json
{ "data": { "relationships": { "foo": { "data": { "id": 1, "type": "foos" } } } }
```

As it's working fine on 0.10.5, it seems to be a regression of 0.10.6

**Comment by bf4** [comment]
@rmachielse Thanks for the bug report. Would you be interested in making a PR to fix this? I know what the problem is, but have limited bandwidth.

**Comment by rmachielse** [comment]
@bf4 hmm so that logic would need to be reverted entirely? As you cannot know upfront if a custom type is used or not

**Comment by bf4** [comment]
> As you cannot know upfront if a custom type is used or not

well, since the serializer specifies the type, and prior to the changes, the serializer was being inferred from the (associated) object, your options are 

1. specify the type on the association
1. specify the serializer on the association (my preference)
1. make the belongs_to behavior that avoids instantiating the associated object (just to look up the serializer) opt-in or opt-out. (This was added to reduce db queries)

**Comment by rmachielse** [comment]
I understand options 1 and 2 but I do not prefer those as it involves some duplication. I assume it is a bug anyway, as this is a breaking change in a patch release. When it comes to option 3, wouldn't it be possible to instantiate the serializer without loading the associated object?

**Comment by sdornan** [comment]
This also seems to happen with `id` on relationships. Even if you specify a serializer that defines an `id` method on the relationship, it doesn't seem to honor that method in the `relationships` block, but does in the `included` block.

Example:

```
...
"relationships" =>
{
    "brand"=>{"data"=>{"id"=> 3, "type"=>"brands"}},
},
"included" =>
[
    {"id"=>"ec949ba7-edee-4d18-a7d8-a3cd351930a9", "type"=>"brands", "attributes"=>{"name"=>"Hyatt", "parent_guid"=>nil, "kind"=>"chain", "category"=>"distinct", "tier"=>"luxury", "image_url"=>nil, "children"=>[]}, "relationships"=>{"promotions"=>{"data"=>[]}}},
]
...
```

**Comment by bf4** [comment]
> I understand options 1 and 2 but I do not prefer those as it involves some duplication. 

I'm not sure I see any duplication. 

> I assume it is a bug anyway, as this is a breaking change in a patch release. 

Yes, I wrote above that it's a bug, when I asked if you wanted to help fix it.

> When it comes to option 3, wouldn't it be possible to instantiate the serializer without loading the associated object?

That's actually not what happens. (I described this in https://github.com/rails-api/active_model_serializers/issues/2125#issuecomment-299874388 ). If it's not told what the associated serializer is, it looks it up from the object.  it *could* look it up from the association type...

**Comment by rmachielse** [comment]
Well I just consider putting the type on the serializer as well as on the relevant relations a duplication. I would by happy to help out, but I am not familiar with the code. 

Would it be an idea to revert [this part](https://github.com/rails-api/active_model_serializers/commit/6e4152851554b33d8cdf607fe4e175dd1aafd1c7#diff-4018ead6cc146fb3c4f9ef621695d850) and let the `ResourceIdentifier` work without using `association.object`? I assume the assocation is loaded as soon as you touch the object?

**Comment by bf4** [comment]
For a PR:

I wouldn't revert the change, but make that code path opt-in. Like `config.belongs_to_uses_id_on_self = true`, maybe? I'd release a 0.10.7 with that bugfix.

Aside:

I really strongly disagree that

```ruby
belongs_to :foo, serializer: FooSerializer
```

is duplication. It's the difference between explicit and implicit. Duplication of typing != duplication of knowledge.  This is particularly true where model->serializer is a one-to-many relationship.  You might want:

```ruby
belongs_to :foo, serializer: AssociatedFooSerializer
```

**Comment by bf4** [comment]
@sdornan re: your comment above, would you open a new issue, please? Thanks :)

**Comment by rmachielse** [comment]
Right, I can agree on that. However when using the type option I feel it's literally repeating the logic more then adding knowledge.

In any case, do you intend to remove the implicit way entirely?

I can certainly pr a config option. Although it looks to me like it can also be resolved without additional config, combining best of both worlds. Or don't you think that will work?

**Comment by MidnightWonderer** [comment]
Since it seems that the code for belongs to association will get rewrite or revert.  
I should mention another part broken with the current behavior.  
I saw the same thing that @sdornan mentioned and also the type inflection configuration is not respected.  
  
Because types come directly from the reflection itself.  
So the inflection inside
```ruby
ResourceIdentifier.type_for
```
does not kick in
```ruby
def self.type_for(class_name, serializer_type = nil, transform_options = {})
  if serializer_type
    raw_type = serializer_type
  else
    inflection =
      if ActiveModelSerializers.config.jsonapi_resource_type == :singular
        :singularize
      else
        :pluralize
      end

    raw_type = class_name.underscore
    raw_type = ActiveSupport::Inflector.public_send(inflection, raw_type)
    raw_type
      .gsub!('/'.freeze, ActiveModelSerializers.config.jsonapi_namespace_separator)
    raw_type
  end
  JsonApi.send(:transform_key_casing!, raw_type, transform_options)
end
```
If I set the 
```ruby
ActiveModelSerializers.config.jsonapi_resource_type = :singular
```
I will get singular type in included data but plural type in relationships block.

**Comment by bf4** [comment]
@MidnightWonderer It does appear to me to be a separate issue. AMS just has too many untested code paths right now.  I'm going to be deprecating master soon, so PR's against 0-10-stable are preferred at this time :)

**Comment by cassidycodes** [comment]
@bf4 Apologies on getting to this so late. I was hit with a cold after our chat last week. Still recovering.

Is this what you had in mind when you said `config.belongs_to_uses_id_on_self = true`? 

https://github.com/cassidycodes/active_model_serializers/commit/f1aa9f7156910c868f54b96807e7e9b90c7507dc

I'm not sure I understand the wording of your suggestion for the config. Doesn't it mean the same thing as `association.belongs_to?` since that method checks `BelongsToReflection#foreign_key_on`?

**Comment by bf4** [comment]
@cassidycodes I commented on our commit. Looks good.

**Comment by cassidycodes** [comment]
@bf4 Thanks! Made a PR, let me know if you want any changes.
